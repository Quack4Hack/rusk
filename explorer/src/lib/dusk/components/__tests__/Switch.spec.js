import { afterEach, describe, expect, it, vi } from "vitest";
import { cleanup, fireEvent, render } from "@testing-library/svelte";

import { Switch } from "..";

describe("Switch", () => {
  const baseProps = {};

  const baseOptions = {
    props: baseProps,
    target: document.body,
  };

  afterEach(cleanup);

  it('should render the "Switch" component with a default tab index of `0`', async () => {
    const { container, rerender } = render(Switch, baseOptions);

    expect(container.firstChild).toMatchSnapshot();

    await rerender({ ...baseProps, value: true });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should use the received tab index", () => {
    const props = {
      ...baseProps,
      tabindex: 5,
    };
    const { container } = render(Switch, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render the component in a disabled status with a tabindex of `-1`", async () => {
    const props = {
      ...baseProps,
      disabled: true,
      tabindex: 5,
    };
    const { container, rerender } = render(Switch, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();

    await rerender({ ...props, value: true });

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should pass additional class names and attributes to the root element", () => {
    const props = {
      ...baseProps,
      className: "foo bar",
      id: "some-id",
    };
    const { container } = render(Switch, { ...baseOptions, props });

    expect(container.firstChild).toMatchSnapshot();
  });

  describe("Event handlers", () => {
    it("should dispatch a `change` event when the switch is clicked", async () => {
      const { component, getByRole } = render(Switch, baseOptions);
      const switchElement = getByRole("switch");
      const handler = vi.fn();

      component.$on("change", handler);

      await fireEvent.click(switchElement);
      await fireEvent.click(switchElement);

      expect(handler).toHaveBeenCalledTimes(2);
      expect(handler).toHaveBeenNthCalledWith(
        1,
        expect.objectContaining({ detail: true })
      );
      expect(handler).toHaveBeenNthCalledWith(
        2,
        expect.objectContaining({ detail: false })
      );
    });

    it("should dispatch a `change` event when the user presses space on the switch", async () => {
      const { component, getByRole } = render(Switch, baseOptions);
      const switchElement = getByRole("switch");
      const handler = vi.fn();

      component.$on("change", handler);

      await fireEvent.keyDown(switchElement, { key: " " });
      await fireEvent.keyDown(switchElement, { key: " " });

      expect(handler).toHaveBeenCalledTimes(2);
      expect(handler).toHaveBeenNthCalledWith(
        1,
        expect.objectContaining({ detail: true })
      );
      expect(handler).toHaveBeenNthCalledWith(
        2,
        expect.objectContaining({ detail: false })
      );
    });

    it("should not dispatch an event if the user presses another key", async () => {
      const { component, getByRole } = render(Switch, baseOptions);
      const switchElement = getByRole("switch");
      const handler = vi.fn();

      component.$on("change", handler);

      await fireEvent.keyDown(switchElement, { key: "Enter" });
      await fireEvent.keyDown(switchElement, { key: "a" });

      expect(handler).not.toHaveBeenCalled();
    });

    it("should not dispatch an event if the switch is disabled", async () => {
      const props = {
        ...baseProps,
        disabled: true,
      };
      const { component, getByRole } = render(Switch, {
        ...baseOptions,
        props,
      });
      const switchElement = getByRole("switch");
      const handler = vi.fn();

      component.$on("change", handler);

      await fireEvent.click(switchElement);
      await fireEvent.keyDown(switchElement, { key: " " });

      expect(handler).not.toHaveBeenCalled();
    });
  });
});
