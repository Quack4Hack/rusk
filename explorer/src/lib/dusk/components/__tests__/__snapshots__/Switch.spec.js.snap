// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch > should pass additional class names and attributes to the root element 1`] = `
<div
  aria-checked="false"
  aria-disabled="false"
  class="dusk-switch foo bar"
  id="some-id"
  role="switch"
  tabindex="0"
/>
`;

exports[`Switch > should render the "Switch" component with a default tab index of \`0\` 1`] = `
<div
  aria-checked="false"
  aria-disabled="false"
  class="dusk-switch"
  role="switch"
  tabindex="0"
/>
`;

exports[`Switch > should render the "Switch" component with a default tab index of \`0\` 2`] = `
<div
  aria-checked="true"
  aria-disabled="false"
  class="dusk-switch"
  role="switch"
  tabindex="0"
/>
`;

exports[`Switch > should render the component in a disabled status with a tabindex of \`-1\` 1`] = `
<div
  aria-checked="false"
  aria-disabled="true"
  class="dusk-switch"
  role="switch"
  tabindex="-1"
/>
`;

exports[`Switch > should render the component in a disabled status with a tabindex of \`-1\` 2`] = `
<div
  aria-checked="true"
  aria-disabled="true"
  class="dusk-switch"
  role="switch"
  tabindex="-1"
/>
`;

exports[`Switch > should use the received tab index 1`] = `
<div
  aria-checked="false"
  aria-disabled="false"
  class="dusk-switch"
  role="switch"
  tabindex="5"
/>
`;
